src/marqo/marqo_url_resolver.py

```diff
@@ -0,0 +1,38 @@
+import time
+import requests
+
+from marqo.errors import MarqoIndexNotFoundError, MarqoIndexNotReadyError
+
+
+class MarqoUrlResolver:
+    def __init__(self, api_key=None, expiration_time: int = 15):
+        """ URL Resolver is a cache for urls that are resolved to their respective indices only for marqo cloud. """
+        self.timestamp = time.time() - expiration_time
+        self._urls_mapping = {"READY": {}, "CREATING": {}}
+        self.api_key = api_key
+        self.expiration_time = expiration_time
+
+    def refresh_urls_if_needed(self, index_name):
+        if index_name not in self._urls_mapping['READY'] and time.time() - self.timestamp > self.expiration_time:
+            # fast refresh to catch if index was created
+            self._refresh_urls()
+        if index_name in self._urls_mapping['READY'] and time.time() - self.timestamp > 360:
+            # slow refresh in case index was deleted
+            self._refresh_urls()
```

Comment:
Consider adding a timeout here, for the slow case (we don't want to hold up search/add docs requests if the control plane is acting badly). Perhaps 3 seconds